File: H16/t16_01_e2303.cpp
#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>

using namespace std;

class Node{
public:
    bool end;
    unordered_map<char, Node*> child;
    Node() : end(false) {}
};

class Tree{
public:
    Node* root;
    Tree() {
        root = new Node();
    }

    bool insert(const string &num) {
        Node* node = root;
        for (char dig: num) {
            if (node->end)
                return false;
            if (node->child.find(dig) == node->child.end())
                node->child[dig] = new Node();
            node = node->child[dig];
        }
        if (node->end || !node->child.empty())
            return false;
        node->end = true;
        return true;
    }
};

string ch(const vector<string> &nums) {
    Tree tree;
    for (const string &num : nums) {
        if (!tree.insert(num))
            return "NO";
    }
    return "YES";
}

int main() {
    int t;
    cin >> t;
    while(t--) {
        int n;
        cin >> n;
        vector<string> nums(n);
        for (int i=0; i<n; i++) {
            cin >> nums[i];
        }
        cout << ch(nums) << endl;
    }
    return 0;
}

File: H16/t16_01_e2303.py
class Node:
    def __init__(self):
        self.children = {}
        self.is_end = False

class Tree:
    def __init__(self):
        self.root = Node()
    def insert(self, number):
        node = self.root
        for digit in number:
            if node.is_end:
                return False
            if digit not in node.children:
                node.children[digit] = Node()
            node = node.children[digit]
        if node.is_end or node.children:
            return False
        node.is_end = True
        return True

def check_phone_list(phone_numbers):
    trie = Tree()
    for number in phone_numbers:
        if not trie.insert(number):
            return "NO"
    return "YES"

t = int(input().strip())
for _ in range(t):
    n = int(input().strip())
    phone_numbers = [input().strip() for _ in range(n)]
    print(check_phone_list(phone_numbers))